packages/components/tests/lib/controllers/app-controller.test.js

```diff
@@ -1,839 +1,1389 @@
+  it('AppController - getWordUsageExamplesQueryParams: should null if word usage examples are disabled', async () => {
+    const textSelector = {
+      languageID: Constants.LANG_LATIN
+    }
+    appC = AppController.jestCreate(uiState)
+    await appC.init()
+    await appC.activate()
+    appC.api.settings.getFeatureOptions().items.wordUsageExamplesON.currentValue = 'onDemand'
+    appC.api.settings.getFeatureOptions().items.enableWordUsageExamples.currentValue = false
+    appC.getWordUsageExamplesQueryParams(textSelector)
+    expect(appC.getWordUsageExamplesQueryParams(textSelector)).toBeNull()
   })
 
-  it('24 AppController - panel methods - setPositionTo, attachToLeft, attachToRight', () => {
-    uiC.panel.setPositionTo('right')
-
-    expect(uiC.panel.options.items.panelPosition.currentValue).toEqual('right')
-
-    uiC.panel.attachToLeft()
-    expect(uiC.panel.options.items.panelPosition.currentValue).toEqual('left')
-
-    uiC.panel.attachToRight()
-    expect(uiC.panel.options.items.panelPosition.currentValue).toEqual('right')
+  it('AppController - handleEscapeKey: should close the UI if the UI is active', async () => {
+    const event = {}
+    const nativeEvent = {
+      keyCode: 27
+    }
+    const uiState = BaseTestHelp.createUIState({ isActive: () => true })
+    appC = AppController.jestCreate(uiState)
+    await appC.init()
+    await appC.activate()
+    const closeUISpy = jest.spyOn(appC.api.ui, 'closeUI')
+    appC.handleEscapeKey(event, nativeEvent)
+    expect(closeUISpy).toBeCalledTimes(1)
   })
 
-  it('25 AppController - panel methods - isOpen, toggle, requestGrammar, showMessage, appendMessage, clearMessages', () => {
-    uiC.panel.open = jest.fn(() => { })
-    uiC.panel.close = jest.fn(() => { })
-
-    uiC.state.isPanelOpen = () => true
-
-    expect(uiC.panel.isOpen()).toBeTruthy()
-
-    uiC.panel.toggle()
-
-    uiC.state.isPanelOpen = () => false
-
-    expect(uiC.panel.isOpen()).toBeFalsy()
-
-    uiC.panel.toggle()
-
-    expect(uiC.panel.open).toHaveBeenCalledTimes(1)
-    expect(uiC.panel.close).toHaveBeenCalledTimes(1)
-
-    ResourceQuery.create = jest.fn(() => { return { getData: () => {} } })
-
-    uiC.panel.requestGrammar()
-    expect(ResourceQuery.create).toHaveBeenCalled()
-
-    uiC.panel.panelData.messages = []
-    uiC.panel.showMessage('fooMessage')
-    expect(uiC.panel.panelData.messages).toEqual([ 'fooMessage' ])
-
-    uiC.panel.appendMessage('fooMessage1')
-    expect(uiC.panel.panelData.messages).toEqual([ 'fooMessage', 'fooMessage1' ])
-
-    uiC.panel.clearMessages()
-    expect(uiC.panel.panelData.messages).toEqual([])
+  it('AppController - handleEscapeKey: should do nothing if the UI is inactive', async () => {
+    const event = {}
+    const nativeEvent = {
+      keyCode: 27
+    }
+    const uiState = BaseTestHelp.createUIState({ isActive: () => false })
+    appC = AppController.jestCreate(uiState)
+    await appC.init()
+    await appC.activate()
```

Comment:
I'm confused about why this test works.  Shouldn't appC.activate() cause the uiState to be active?

Reply:
> Shouldn't appC.activate() cause the uiState to be active?

In the real objects, yes. `appC.activate()` would call `activate()` of the UI controller. We, however, use the mock of the `uiState` and the `activate()` method, when called upon it, will do nothing. Even though if it would be the real object, we redefine its `isActive` method so that it will always return false: `const uiState = BaseTestHelp.createUIState({ isActive: () => false })`. Because of that, the UI would be identified as inactive.

I agree it's a hacky way to test, but I think it's probably OK considering that the UI state is a mock, not a fully-functional object. I think we should test with the real version of the `uiState` ideally, but considering complexities of that and the fact that we're planning to change the way options are stored and handled maybe we can postpone that for later?

I can add a comment explaining the situation.

What do you think?

Reply:
sure, a comment would be helpful!